### PR TITLE
feature/s21_strncpy

### LINKED ISSUE #26: `s21_strncpy` - Copies up to `n` characters from `src` to `dest`.
## Description
Implement the `s21_strncpy()` function to copy up to `n` characters from `src` to `dest`, matching the behavior of the standard `strncpy()` function (POSIX.1-2017).

## Requirements

- [ ] Follow Google C++ Style Guide  
- [ ] 100% test coverage (via Check library)  
- [ ] No memory leaks  
- [ ] Documentation in header file  
- [ ] Match standard library behavior exactly including:
   - Padding with null bytes when src < n  
   - No null-termination when src ≥ n  
   - Return value equals dest  
   - Handle overlapping memory regions  

## Test Cases Needed
- Basic string copy
- Partial copy (n < src length)
- Full copy with padding (n > src length) (probably bad idea)
- Empty source string
- Copy with n=0
- ASCII compatibility
- Destination buffer exactly fits source
- Overlapping memory regions

## Files to Modify
- `src/s21_strncpy.c` (implementation)
- `src/s21_string.h` (declaration)
- `tests/test_strncpy.c` (unit tests)
- `tests/suites.h`
- `tests/main.c`

### PR DESCRIPTION
**Pull request by DarkDomian**
## Changes | Closed #26 
✨ Added `s21_strncpy()` implementation:
- Copies exactly `n` bytes from `src` to `dest`
- Properly handles padding with null bytes
- Returns `dest` pointer as specified
- Matches standard library edge cases

🧪 Added comprehensive test coverage:
- 15 test cases covering all edge conditions
- Memory comparison tests
- Buffer overflow checks
- Padding verification

📚 Updated documentation:
- Added detailed function description in `s21_string.h`
- Specified padding and termination behavior

## Verification

- [ ] All tests pass (Check library)  
- [ ] 100% line coverage (gcov)  
- [ ] Valgrind-clean (no memory leaks)  
- [ ] Clang-format compliant 

## Example Usage
```c
char dest[10];
s21_strncpy(dest, "hello", 10);
// Copies "hello" + 5 null bytes
```

## Testing
```bash
make style-check
make strncpy.test
```

### CODE REVIEW

File: tests/test_strncpy.c [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                   s21_strncpy(dest2, src, BUF_SIZE), BUF_SIZE);
+}
+END_TEST
+
+START_TEST(test_strncpy_overlap) {
+  char buffer1[] = "abcdefghij";
+  char buffer2[] = "abcdefghij";
+
+  s21_strncpy(buffer1 + 3, buffer1, 4);
+  strncpy(buffer2 + 3, buffer2, 4);
```
Comment (gashturik) [inline tests/test_strncpy.c]: raise the cppcheck error:
```c
tests/test_strncpy.c:53:3: error: Overlapping read/write in strncpy() is undefined behavior [overlappingWriteFunction]
  strncpy(buffer2 + 3, buffer2, 4);
  ^
```

**Review by gashturik — COMMENTED** [review]
A lot of errors like `undefined reference to`  `any check.h function` now. And finally
```
collect2: error: ld returned 1 exit status
make: *** [Makefile:85: test] Ошибка 1
```

File: tests/test_strncpy.c (hunk 1 above)
Comment (ParthenGithub) [inline tests/test_strncpy.c]: +1
```c
Checking tests/test_strncpy.c ...
tests/test_strncpy.c:53:3: error: Overlapping read/write in strncpy() is undefined behavior [overlappingWriteFunction]
  strncpy(buffer2 + 3, buffer2, 4);
```

File: src/s21_strncpy.c [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#include "../headers/s21_string.h"
+
+char *s21_strncpy(char *dest, const char *src, s21_size_t n) {
```
Comment (ParthenGithub) [inline src/s21_strncpy.c:3]: Valgrind didn't pass
```c
==472391== Source and destination overlap in strncpy(0x1ffefff0ac, 0x1ffefff0a9, 4)
==472391==    at 0x4E0E388: strncpy (vg_replace_strmem.c:607)
==472391==    by 0x10EB9C: test_strncpy_overlap_fn (in /home/rhydonte/Desktop/Projects/STRING/test)
==472391==    by 0x1154B8: tcase_run_tfun_nofork (in /home/rhydonte/Desktop/Projects/STRING/test)
==472391==    by 0x115927: srunner_run_tagged (in /home/rhydonte/Desktop/Projects/STRING/test)
==472391==    by 0x10BB10: main (in /home/rhydonte/Desktop/Projects/STRING/test)
==472391== 
100%: Checks: 26, Failures: 0, Errors: 0
==472391== 
==472391== HEAP SUMMARY:
==472391==     in use at exit: 18 bytes in 1 blocks
==472391==   total heap usage: 3,291 allocs, 3,290 frees, 418,551 bytes allocated
==472391== 
==472391== LEAK SUMMARY:
==472391==    definitely lost: 0 bytes in 0 blocks
==472391==    indirectly lost: 0 bytes in 0 blocks
==472391==      possibly lost: 0 bytes in 0 blocks
==472391==    still reachable: 18 bytes in 1 blocks
==472391==         suppressed: 0 bytes in 0 blocks
==472391== Reachable blocks (those to which a pointer was found) are not shown.
==472391== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==472391== 
==472391== For lists of detected and suppressed errors, rerun with: -s
==472391== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

**Review by ParthenGithub — CHANGES_REQUESTED** [review]

**Review by ParthenGithub — APPROVED** [review]

File: tests/test_strncpy.c (hunk 1 above)
Comment (gashturik) [reply tests/test_strncpy.c]: `sudo apt install pkg-config` solve the problem

**Review by Amfir — APPROVED** [review]

File: src/s21_strncpy.c (hunk 2 above)
Comment (DarkDomian) [reply src/s21_strncpy.c:3]: Resolved by removing overlapping call of `strnpcy()`

File: tests/test_strncpy.c (hunk 1 above)
Comment (DarkDomian) [reply tests/test_strncpy.c]: Relosved by removing overlapping call of `strncpy()`

Comment (DarkDomian) [reply tests/test_strncpy.c]: Resolved by removing overlapping call of `strncpy()`
